fix flake on Conformance test should ResourceQuota apply changes to a…

aojea opened the pull request:
… resourcequota status

The test flakes when trying to verify that, after modifying the ResourceQuota status, the controller reconciles the value to the spec. However, this reconcilation only happens periodically and the default value of this loop is 12h, so is hard to test this reliably in all clusters.



/kind failing-test
/kind flake

Fixes #121911

#### Special notes for your reviewer:

See https://github.com/kubernetes/kubernetes/issues/121911#issuecomment-1817473968

```release-note
NONE
```

/priority critical-urgent

ping @kubernetes/release-managers 
/assign @liggitt @jpbetz

aojea commented on test/e2e/apimachinery/resource_quota.go:
```diff
@@ -1186,21 +1186,11 @@ var _ = SIGDescribe("ResourceQuota", func() {
 		})
 		framework.ExpectNoError(err, "failed to locate ResourceQuota %q in namespace %q", patchedResourceQuota.Name, ns)
 
-		err = wait.PollUntilContextTimeout(ctx, 5*time.Second, 5*time.Minute, true, func(ctx context.Context) (bool, error) {
```
gce tests use a 3 mins value for min-sync period, kind uses kubeadm and uses a 12h value for the min-sync period.
My reading of the code is that the resource-quota-controller ignore changes to the status and reconciles it every ResyncPeriod, and this is a random value obtained from the kube-controller-manager flag min-sync-period and  is a value between min-sync-period and 2*min-sync-period.

neolit123 replied:
cc @Riaankl 
https://github.com/kubernetes/kubernetes/commit/168fa73b60b625ed1e69f2cd62f850ba46a24efc
(conformance promotion)

neolit123 replied:
if the resync period averages at 18h on kubeadm jobs and this part of the test is supposed to fail on the stock defaults, why is the test 99% green on kubeadm jobs?
https://testgrid.k8s.io/sig-cluster-lifecycle-kubeadm#kubeadm-kinder-latest

aojea replied:
that part I don't understand it too, that is why I asked for reviewers that know better the resource controller, but you can follow the codepath in https://github.com/kubernetes/kubernetes/issues/121911#issuecomment-1817473968

aojea replied:
/hold
we should be able to answer this question, there is a `pkg/controller/resourcequota/resource_quota_monitor.go` that I don't have clear what is doing

aojea replied:
/hold cancel
@neolit123 now I think I got it right, thanks for the great question

aojea replied:
updated the PR, it turns out the resyn period i other option https://github.com/kubernetes/kubernetes/issues/121911#issuecomment-1817593892

aojea committed:
fix flake on conformance e2e test ResourceQuota controller should apply changes to a resourcequota status

The e2e test patch the status of a ResourceQuota resources and tries to
verify the controller reset its status, however, the controller ignores
the updates and only reconcile the objects every a predefined interval,
by default 5 minutes.

Since the test polls for 5 minutes, there are some edge cases that the
time to reconcile the object by the reconcile loop is greater than 5
minutes failing the test.

To take into account the time to reconcile the objects and the reconcile
loop period, we increase by one minute the poll loop.

Change-Id: I30f7fda36cdfb47c543b5b2b120e39f7d6c2442d

neolit123 commented on test/e2e/apimachinery/resource_quota.go:
```diff
@@ -1186,21 +1186,36 @@ var _ = SIGDescribe("ResourceQuota", func() {
-			if apiequality.Semantic.DeepEqual(resourceQuotaResult.Spec.Hard.Cpu(), resourceQuotaResult.Status.Hard.Cpu()) {
-				gomega.Expect(*resourceQuotaResult.Status.Hard.Cpu()).To(gomega.Equal(resource.MustParse("1")), "Hard cpu value for ResourceQuota %q is %s not 1.", repatchedResourceQuota.Name, repatchedResourceQuota.Status.Hard.Cpu().String())
-				gomega.Expect(*resourceQuotaResult.Status.Hard.Memory()).To(gomega.Equal(resource.MustParse("1Gi")), "Hard memory value for ResourceQuota %q is %s not 1Gi.", repatchedResourceQuota.Name, repatchedResourceQuota.Status.Hard.Memory().String())
+			if *resourceQuotaResult.Spec.Hard.Cpu() != *resourceQuotaResult.Status.Hard.Cpu() {
```
was the semantic equality causing a problem for these Quantity structs from Cpu()?
https://github.com/kubernetes/apimachinery/blob/master/pkg/api/equality/semantic.go#L29
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/api/resource/quantity.go#L100
 
comparison by value should work fine, but it would also include the format string.

aojea replied:
this code is much more easier to read and is much simpler

neolit123 reviewed (commented):
the assertion refactor and timeout change LGTM

neolit123 commented on test/e2e/apimachinery/resource_quota.go:
```diff
@@ -1186,21 +1186,36 @@ var _ = SIGDescribe("ResourceQuota", func() {
 		})
-		framework.ExpectNoError(err)
+		if err != nil {
+			framework.Failf("Error waiting for ResourceQuota %q to reset its Status", patchedResourceQuota.Name, err)
```
```suggestion
			framework.Failf("Error waiting for ResourceQuota %q to reset its Status: %v", patchedResourceQuota.Name, err)
```

jpbetz commented:
/approve

aojea commented:
/retest

known flake https://github.com/kubernetes/kubernetes/issues/121865

ping @kubernetes/release-managers 

this just needs lgtm and milestone

neolit123 reviewed (commented):
/lgtm
/milestone v1.29
the diff SGTM, as is.

/hold
drop the hold if you don't want to wait for release-managers or more review.

jeremyrickard reviewed (commented):
/lgtm

will leave dropping the hold to @Priyankasaggu11929 and team for the release.

dims commented:
/approve
/lgtm

Priyankasaggu11929 commented:
/approve
(from 1.29 Release Team)

/unhold

liggitt commented on test/e2e/apimachinery/resource_quota.go:
```diff
@@ -1186,21 +1186,36 @@ var _ = SIGDescribe("ResourceQuota", func() {
+		// Ref: https://issues.k8s.io/121911
+		err = wait.PollUntilContextTimeout(ctx, 5*time.Second, 6*time.Minute, true, func(ctx context.Context) (bool, error) {
 			resourceQuotaResult, err := rqClient.Get(ctx, rqName, metav1.GetOptions{})
-			framework.ExpectNoError(err)
```
we don't expect the get to error, right? when would that be expected / ok?

aojea replied:
this is inside the poll loop,  why should we fail hard here? usually we always retry in these cases

liggitt replied:
The poll is to give the reconciler time to make the status value what we want. An unstable API server here is unexpected

aojea replied:
I  stand corrected, after a quick grep  I have to fully agree with you, I expected to see more retries on other tests, but it seems the norm is to fail here

aojea replied:
https://github.com/kubernetes/kubernetes/pull/121972

neolit123 replied:
if this GET API call must be one shot and not fail then this:
```go
resourceQuotaResult, err := rqClient.Get(ctx, rqName, metav1.GetOptions{})
framework.ExpectNoError(err)
```
must be outside of the poll loop.

aojea replied:
but you have to try to get everytime, or you don't get the new data, returning an error makes the same efect and avoid panicing inside https://github.com/kubernetes/kubernetes/pull/121972

neolit123 replied:
yes, and that's why this part of this PR, LGTM.

the change in https://github.com/kubernetes/kubernetes/pull/121972 is an improvement where the overall poll will know about the reported GET errors. that LGTM as well.

liggitt commented on test/e2e/apimachinery/resource_quota.go:
```diff
@@ -1186,21 +1186,36 @@ var _ = SIGDescribe("ResourceQuota", func() {
-			if apiequality.Semantic.DeepEqual(resourceQuotaResult.Spec.Hard.Cpu(), resourceQuotaResult.Status.Hard.Cpu()) {
-				gomega.Expect(*resourceQuotaResult.Status.Hard.Cpu()).To(gomega.Equal(resource.MustParse("1")), "Hard cpu value for ResourceQuota %q is %s not 1.", repatchedResourceQuota.Name, repatchedResourceQuota.Status.Hard.Cpu().String())
-				gomega.Expect(*resourceQuotaResult.Status.Hard.Memory()).To(gomega.Equal(resource.MustParse("1Gi")), "Hard memory value for ResourceQuota %q is %s not 1Gi.", repatchedResourceQuota.Name, repatchedResourceQuota.Status.Hard.Memory().String())
+			if *resourceQuotaResult.Spec.Hard.Cpu() == *resourceQuotaResult.Status.Hard.Cpu() {
```
dereferencing and comparing rather than using Quantity#Equal seems fragile

aojea replied:
that is something I was wondering about, when I read the library I saw the euqality function but then in the same test I saw it was comparing the values, just for learning, what is the difference?

neolit123 replied:
my earlier comment:
https://github.com/kubernetes/kubernetes/pull/121951#discussion_r1398261216

the comparison by value is fine for now, but if the struct introduces pointer objects it would fail.

```go
package main

import "fmt"

type bar struct { i int }
type foo struct { b *bar }

func main() {
	a := &foo{b: &bar{i: 10}}
	b := &foo{b: &bar{i: 10}}
	fmt.Printf("a == b = %v", *a == *b) // a == b = false
}
```

liggitt replied:
.Equal compares equality of the values, which may or may not be identical to struct equality, depending on normalization and whether the string rendering is cached

```go
func TestCPUComparison(t *testing.T) {
	l1 := ResourceList{"cpu": resource.MustParse("1m")}
	l2 := ResourceList{"cpu": resource.MustParse(".001")}

	cpu1 := l1.Cpu()
	cpu2 := l2.Cpu()

	t.Logf("%#v", cpu1)
	t.Logf("%#v", cpu2)

	if cpu1.Equal(*cpu2) {
		t.Log("passed Equal()")
	} else {
		t.Fatal("failed Equal()")
	}

	if *cpu1 == *cpu2 {
		t.Log("passed ==")
	} else {
		t.Fatal("failed ==")
	}
}
```

outputs

```sh
=== RUN   TestCPUComparison
    resource_test.go:16: &resource.Quantity{i:resource.int64Amount{value:1, scale:-3}, d:resource.infDecAmount{Dec:(*inf.Dec)(nil)}, s:"1m", Format:"DecimalSI"}
    resource_test.go:17: &resource.Quantity{i:resource.int64Amount{value:1, scale:-3}, d:resource.infDecAmount{Dec:(*inf.Dec)(nil)}, s:"", Format:"DecimalSI"}
    resource_test.go:20: passed Equal()
    resource_test.go:28: failed ==
--- FAIL: TestCPUComparison (0.00s)
```

liggitt commented:
did this flake just start? do we know why? this test hasn't changed in ~6 months, right?

aojea commented:
> did this flake just start? do we know why? this test hasn't changed in ~6 months, right?

my guess is that it was failing always, the analyzed failure seems correct https://github.com/kubernetes/kubernetes/issues/121911#issuecomment-1817473968 , it reconcile every 5 minutes and if the loop poll starts about that time it will not be able to sync the status on time
